Sources/SpeziLLMLocal/LLMLocalSession+Generate.swift

```diff
@@ -53,78 +55,104 @@ extension LLMLocalSession {
+            return
+        }
+    }
+    
+    private func prepareModelInput(messages: [[String: String]], modelContainer: ModelContainer) async throws -> LMInput {
+        try await modelContainer.perform { modelContext in
+            if let chatTemplate = self.schema.parameters.chatTemplate {
+                let tokens = try modelContext.tokenizer.applyChatTemplate(messages: messages, chatTemplate: chatTemplate)
+                return LMInput(text: .init(tokens: MLXArray(tokens)))
+            } else {
+                return try await modelContext.processor.prepare(input: .init(messages: messages))
+            }
+        }
+    }
+    
+    private func processTokens(
+        _ tokens: [Int],
+        modelContext: ModelContext,
+        continuation: AsyncThrowingStream<String, any Error>.Continuation
+    ) -> GenerateDisposition {
+        if Task.isCancelled {
+            return .stop
+        }
+        
+        if tokens.count >= self.schema.parameters.maxOutputLength {
+            Self.logger.debug("SpeziLLMLocal: Max output length exceeded.")
+            return .stop
+        }
+        
+        if tokens.count.isMultiple(of: schema.parameters.displayEveryNTokens) {
+            let lastTokens = Array(tokens.suffix(schema.parameters.displayEveryNTokens))
+            let text = modelContext.tokenizer.decode(tokens: lastTokens)
+            
+            Self.logger.debug("SpeziLLMLocal: Yielded token: \(text, privacy: .public)")
             continuation.yield(text)
             
             if schema.injectIntoContext {
                 Task { @MainActor in
                     context.append(assistantOutput: text)
-                    context.completeAssistantStreaming()
                 }
             }
-            
-            return result
         }
         
-        Self.logger.debug(
-            """
-            SpeziLLMLocal:
-            Prompt Tokens per second: \(result.promptTokensPerSecond, privacy: .public)
-            Generation tokens per second: \(result.tokensPerSecond, privacy: .public)
-            """
-        )
+        return .more
+    }
+    
+    private func processRemainingTokens(
+        result: GenerateResult,
+        modelContext: ModelContext,
+        continuation: AsyncThrowingStream<String, any Error>.Continuation
+    ) {
+        // Yielding every Nth token may result in missing the final tokens.
+        let remainingTokens = result.tokens.count % schema.parameters.displayEveryNTokens
+        let lastTokens = Array(result.tokens.suffix(remainingTokens))
+        let text = modelContext.tokenizer.decode(tokens: lastTokens)
+        continuation.yield(text)
         
-        await MainActor.run {
-            continuation.finish()
-            state = .ready
+        if schema.injectIntoContext {
+            Task { @MainActor in
+                context.append(assistantOutput: text)
+                context.completeAssistantStreaming()
+            }
```

Comment:
Could we make this func `async` and avoid scheduling a Task by using `MainActor.run { ... }`?

Reply:
The closure for `MLXLMCommon.generate`, where this function is called, has to be synchronous `([Int]) -> GenerateDisposition`, so we would still have to schedule a Task either way. Is that better than this?